src/search/plugins/types.cc

```diff
@@ -0,0 +1,345 @@
+#include "types.h"
+
+#include "plugin.h"
+
+#include "../utils/strings.h"
+
+#include <sstream>
+#include <typeindex>
+
+using namespace std;
+
+namespace plugins {
+bool Type::operator!=(const Type &other) const {
+    return !(*this == other);
+}
+
+bool Type::is_basic_type() const {
+    return false;
+}
+
+const type_index &Type::get_basic_type_index() const {
+    ABORT("Used Type::get_basic_type_index on a type that does not support it.");
+}
+
+bool Type::is_feature_type() const {
+    return false;
+}
+
+bool Type::is_predefinable() const {
+    return false;
+}
+
+string Type::get_synopsis() const {
+    ABORT("Used Type::get_synopsis on a type that does not support it.");
+}
+
+bool Type::is_list_type() const {
+    return false;
+}
+
+bool Type::has_nested_type() const {
+    return false;
+}
+
+const Type &Type::get_nested_type() const {
+    ABORT("Used Type::get_nested_type on a type that does not support it.");
+}
+
+bool Type::is_enum_type() const {
+    return false;
+}
+
+int Type::get_enum_index(const string &) const {
+    ABORT("Used Type::get_enum_index on a type that does not support it.");
+}
+
+const EnumInfo &Type::get_documented_enum_values() const {
+    ABORT("Used Type::get_documented_enum_values on a type that does not support it.");
+}
+
+bool Type::is_symbol_type() const {
+    return false;
+}
+
+bool Type::can_convert_into(const Type &other) const {
+    return *this == other;
+}
+
+BasicType::BasicType(type_index type, const string &class_name)
+    : type(type), class_name(class_name) {
+}
+
+bool BasicType::operator==(const Type &other) const {
+    const BasicType *other_ptr = dynamic_cast<const BasicType *>(&other);
+    return other_ptr && type == other_ptr->type;
+}
+
+bool BasicType::is_basic_type() const {
+    return true;
+}
+
+const type_index &BasicType::get_basic_type_index() const {
+    return type;
+}
+
+bool BasicType::can_convert_into(const Type &other) const {
+    return Type::can_convert_into(other)
+           || (other.is_basic_type()
+               && get_basic_type_index() == typeid(int)
+               && other.get_basic_type_index() == typeid(double));
+}
+
+string BasicType::name() const {
+    return class_name;
+}
+
+size_t BasicType::hash() const {
+    return std::hash<type_index>()(type);
```

Comment:
This is problematic: leaving out `std::` does not compile anymore (for some reason the `using namespace std;` is not sufficient). But leaving the `std::` causes our style checks to fail.